src/SystemManager.cc

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "SystemManager.hh"
+
+using namespace ignition;
+using namespace gazebo;
+
+//////////////////////////////////////////////////
+SystemManager::SystemManager(const SystemLoaderPtr &_systemLoader,
+                             EntityComponentManager *_entityCompMgr,
+                             EventManager *_eventMgr)
+  : systemLoader(_systemLoader),
+    entityCompMgr(_entityCompMgr),
+    eventMgr(_eventMgr)
+{
+}
+
+//////////////////////////////////////////////////
+void SystemManager::LoadPlugin(const Entity _entity,
+                               const std::string &_fname,
+                               const std::string &_name,
+                               const sdf::ElementPtr &_sdf)
+{
+  std::optional<SystemPluginPtr> system;
+  {
+    std::lock_guard<std::mutex> lock(this->systemLoaderMutex);
+    system = this->systemLoader->LoadPlugin(_fname, _name, _sdf);
+  }
+
+  // System correctly loaded from library
+  if (system)
+  {
+    this->AddSystem(system.value(), _entity, _sdf);
+    igndbg << "Loaded system [" << _name
+           << "] for entity [" << _entity << "]" << std::endl;
+  }
+}
+
+//////////////////////////////////////////////////
+size_t SystemManager::TotalCount() const
+{
+  std::lock_guard<std::mutex> lock(this->systemsMutex);
+  return this->systems.size() + this->pendingSystems.size();
+}
+
+//////////////////////////////////////////////////
+size_t SystemManager::ActiveCount() const
+{
+  std::lock_guard<std::mutex> lock(this->systemsMutex);
```

Comment:
No need to lock a mutex here, I don't think. See my comment where this mutex is declared.